src/app/introduction/picture-puzzle.component.ts

```diff
@@ -0,0 +1,384 @@
+                    this._stage.globalAlpha = .4;
+                    this._stage.fillStyle = this.PUZZLE_HOVER_TINT;
+                    this._stage.fillRect(this._currentDropPiece.xPos, this._currentDropPiece.yPos, this._pieceWidth, this._pieceHeight);
+                    this._stage.restore();
+                }
+            }
+        }
+        this._stage.save();
+        this._stage.globalAlpha = .6;
+        this._stage.drawImage(this._newImg, this._currentPiece.sx, this._currentPiece.sy, this._pieceWidth, this._pieceHeight, this._mouse.x - (this._pieceWidth / 2), this._mouse.y - (this._pieceHeight / 2), this._pieceWidth, this._pieceHeight);
+        this._stage.restore();
+        this._stage.strokeRect( this._mouse.x - (this._pieceWidth / 2), this._mouse.y - (this._pieceHeight / 2), this._pieceWidth, this._pieceHeight);
+    }
+
+    pieceDropped(e) {
+        this._mousemoveListener();
+        this._mouseupListener();
+        if (this._currentDropPiece != null) {
+            const tmp = {xPos: this._currentPiece.xPos, yPos: this._currentPiece.yPos};
+            this._currentPiece.xPos = this._currentDropPiece.xPos;
+            this._currentPiece.yPos = this._currentDropPiece.yPos;
+            this._currentDropPiece.xPos = tmp.xPos;
+            this._currentDropPiece.yPos = tmp.yPos;
+        }
+        this.resetPuzzleAndCheckWin();
+    }
+
+    resetPuzzleAndCheckWin() {
+        this._stage.clearRect(0, 0, this._puzzleWidth, this._puzzleHeight);
+        let gameWin = true;
+        let i, piece;
+        for (i = 0; i < this._pieces.length; i++) {
+            piece = this._pieces[i];
+            this._stage.drawImage(this._newImg, piece.sx, piece.sy, this._pieceWidth, this._pieceHeight, piece.xPos, piece.yPos, this._pieceWidth, this._pieceHeight);
+            this._stage.strokeRect(piece.xPos, piece.yPos, this._pieceWidth, this._pieceHeight);
+            if (piece.xPos !== piece.sx || piece.yPos !== piece.sy) {
+                gameWin = false;
+            }
+        }
+        if (gameWin) {
+            this.gameOver();
+        }
+    }
+
+    gameOver() {
+        this._dashboardService.updateProgressStatus(this._status).subscribe(response => {});
+        swal(
+            'Good job!',
+            'You completed this activity!',
+            'success'
+        );
+        this._mousedownListener();
+        this._mousemoveListener();
+        this._mouseupListener();
+        this.initPuzzle();
+        this.activityComplete = true;
+    }
+
+    /**
+     * Check progress of user (If the user has already completed the activity or not)
+     */
+    checkProgress() {
+        this._dashboardService.getProgressStatus().subscribe(response => {
+            this._activity = response.activity;
+            this._stage = response.stage;
+            if (this._stage >= 1 && this._activity >= 3) {
+                this.activityComplete = true;
+            }
+        });
+    }
+}
```

Comment:
try to clean this file up slightly, especially reduce the amount of repeated code

good reads: http://www.itiseezee.com/?p=119
this is slightly extreme, but does have few good points & practices